### PR TITLE
chore: update crypto compare with new websocket impl

### LINKED ISSUE #13: Make websocket subscriptions come back to life after a disconnection
Currently if the websockets connection fails for some reason, there are no mechanisms to re-initialise it and we need to restart the service using it to bring it back to life.

We need to improve the existing websocket client to make sure that it resurrects automatically if possible.

It would be great to find a pattern that depends as little as possible on the specifics of the CryptoCompare API, then we can re-use it in other api clients easily.

This should resolve https://github.com/trakx/trakx-tools/issues/381

It should also give us good ideas to move onto the next task: 
- https://github.com/trakx/exchange-api-client/issues/15

and then 
- https://github.com/trakx/market-making/issues/4

### PR DESCRIPTION
**Pull request by JayArrowz**
- Update crypto compare API to new websocket framework
- Depends on https://github.com/trakx/websocket/pull/4 which adds reconnection resurrection 
- Closes #13

### CODE REVIEW

File: src/Trakx.CryptoCompare.ApiClient/Trakx.CryptoCompare.ApiClient.csproj [hunk 1]
```diff
@@ -17,11 +17,12 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.Extensions.Options.ConfigurationExtensions" Version="5.0.0" />
-    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.1" />
     <PackageReference Include="Polly" Version="7.2.1" />
     <PackageReference Include="Serilog" Version="2.10.0" />
     <PackageReference Include="System.Reactive.Linq" Version="5.0.0" />
     <PackageReference Include="Trakx.Utils" Version="0.1.13" />
+    <PackageReference Include="Trakx.Websocket" Version="0.3.31-a4763ee" />
```
Comment (RafaelMeloTrakx) [inline src/Trakx.CryptoCompare.ApiClient/Trakx.CryptoCompare.ApiClient.csproj:25]: In dotnet environment, websocket is usually written with "S" in upper case ("WebSocket"). not a big deal.

File: src/Trakx.CryptoCompare.ApiClient/ServiceConfiguration.cs [hunk 2]
```diff
@@ -31,10 +33,9 @@ public static IServiceCollection AddCryptoCompareClient(this IServiceCollection
         private static void AddCommonDependencies(IServiceCollection services)
         {
             services.AddSingleton<ICryptoCompareClient, CryptoCompareClient>();
-            services.AddTransient<IClientWebsocket, WrappedClientWebsocket>();
```
Comment (RafaelMeloTrakx) [inline src/Trakx.CryptoCompare.ApiClient/ServiceConfiguration.cs:34]: Nice!

File: src/Trakx.CryptoCompare.ApiClient.Websocket.Tests/Unit/CryptoCompareWebsocketHandlerTests.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            var fakeFinexConfig = new CryptoCompareWebsocketConfiguration
+            {
+                Url = "https://www.google.com"
+            };
+
+            _websocketHandler = new CryptoCompareWebsocketHandler(fakeOptions, fakeFinexConfig, fakeWebsocketFactory);
+        }
+
+        [Fact]
+        public async Task AddAsync_should_not_remove_subscription_on_sub_remove()
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.Websocket.Tests/Unit/CryptoCompareWebsocketHandlerTests.cs:45]:
```suggestion
        public async Task AddAsync_should_remove_subscription_on_sub_remove()
```
Or am I confused?

File: src/Trakx.CryptoCompare.ApiClient.Websocket.Tests/Unit/CryptoCompareWebsocketHandlerTests.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+            var topicMessageTask = _websocketHandler.GetTopicMessageStream<FullVolume>(topicSub.Topic).Buffer(TimeSpan.FromSeconds(5), 1)
+                .Select(t => t.FirstOrDefault())
+                .FirstOrDefaultAsync()
+                .ToTask();
+            var fakeMsg = JsonSerializer.Serialize(new FullVolume { Type = "11", Volume = 1, Symbol = "BTC" });
+            _testClient.StreamFakeMessage(ResponseMessage.TextMessage(fakeMsg));
+
+            var topicMessage = await topicMessageTask;
+            topicMessage.Should().NotBeNull();
+
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.Websocket.Tests/Unit/CryptoCompareWebsocketHandlerTests.cs:73]: then should we test the content is as expected while we are there?

File: src/Trakx.CryptoCompare.ApiClient.Websocket/CryptoCompareWebsocketConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using Trakx.Utils.Attributes;
+
+namespace Trakx.CryptoCompare.ApiClient.Websocket
+{
+    public class CryptoCompareWebsocketConfiguration
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.Websocket/CryptoCompareWebsocketConfiguration.cs:5]: let's make it an init only record and suppress the warnings or add a constructor for those props

File: src/Trakx.CryptoCompare.ApiClient.Websocket/Model/CryptoCompareSubscription.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿using System.Collections.Generic;
+using System.Text.Json.Serialization;
+
+namespace Trakx.CryptoCompare.ApiClient.Websocket.Model
+{
+    public class CryptoCompareSubscription
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.Websocket/Model/CryptoCompareSubscription.cs:6]: please suppress the nullable warnings ;)

File: src/Trakx.CryptoCompare.ApiClient.Websocket/Model/HeartBeat.cs [hunk 7]
```diff
@@ -1,10 +1,9 @@
 ﻿using System.Text.Json.Serialization;
 
-namespace Trakx.CryptoCompare.ApiClient.WebSocket.DTOs.Inbound
+namespace Trakx.CryptoCompare.ApiClient.Websocket.Model
 {
     public class HeartBeat : InboundMessageBase
     {
-        internal const string TypeValue = "999";
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.Websocket/Model/HeartBeat.cs:7]: Feels good to remove these I think, I didn't know if it was better to have them implemented in/near the class like that, or in a separate class only in charge of the mapping like you did...

File: src/Trakx.CryptoCompare.ApiClient.Websocket/Model/OrderBookL2.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+﻿using System.Text.Json.Serialization;
+
+namespace Trakx.CryptoCompare.ApiClient.Websocket.Model
+{
+    public class OrderBookL2 : InboundMessageBase
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.Websocket/Model/OrderBookL2.cs:5]: please suppress the nullable warnings ;)

File: src/Trakx.CryptoCompare.ApiClient.Websocket/Model/Welcome.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+using System.Text.Json.Serialization;
+using System.Threading.Tasks;
+
+namespace Trakx.CryptoCompare.ApiClient.Websocket.Model
+{
+    public class Welcome : InboundMessageBase
+    {
+#nullable disable
+        [JsonPropertyName("TYPE")]
+        public string Type { get; set; }
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.Websocket/Model/Welcome.cs:14]: can you remove this prop maybe? cf. warning message

File: src/Trakx.CryptoCompare.ApiClient.sln [hunk 10]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.30709.132
+# Visual Studio Version 17
+VisualStudioVersion = 17.0.31903.59
```
Comment (monsieurleberre) [inline src/Trakx.CryptoCompare.ApiClient.sln:4]: Oooooh 🤗

**Review by monsieurleberre — COMMENTED** [review]
Nice, it looks like the library you built can be smoothly adapted to this case ❤️ well done.
Can we keep a couple of integration tests please though? Just for sanity checking!

Thanks

**Comment by JayArrowz** [comment]
> Nice, it looks like the library you built can be smoothly adapted to this case ❤️ well done. Can we keep a couple of integration tests please though? Just for sanity checking!
> 
> Thanks

yessir! writing integration tests now

**Comment by JayArrowz** [comment]
Added integration tests

File: src/Trakx.CryptoCompare.ApiClient.sln (hunk 10 above)
Comment (JayArrowz) [reply src/Trakx.CryptoCompare.ApiClient.sln:4]: VS x64 ftw

File: src/Trakx.CryptoCompare.ApiClient.Websocket.Tests/Unit/CryptoCompareWebsocketHandlerTests.cs (hunk 4 above)
Comment (JayArrowz) [reply src/Trakx.CryptoCompare.ApiClient.Websocket.Tests/Unit/CryptoCompareWebsocketHandlerTests.cs:76]: Added content test

**Review by monsieurleberre — APPROVED** [review]
thanks!

**Pull request merged**